### PR TITLE
Permission implementation

### PR DESCRIPTION
**Pull request by deye9**

### CODE REVIEW

File: models/src/main/java/com/foodie/FoodieBaseService.java [hunk 1]
```diff
@@ -86,4 +88,30 @@ default Page<T> findAllByDeletedAtIsNull(int page, int size, String sort) {
...
+        for (Field field : fields) {
+            if (!isFieldIgnored(field)) {
+                field.setAccessible(true);
+                try {
+                    Object value = field.get(sourceEntity);
+                    if (value != null) {
+                        field.set(targetEntity, value);
+                    }
+                } catch (IllegalAccessException e) {
+                    e.printStackTrace();
```
Comment (sfagade) [inline models/src/main/java/com/foodie/FoodieBaseService.java:105]: Use logger instead

File: models/src/main/java/com/foodie/FoodieBaseService.java [hunk 2]
```diff
@@ -1,16 +1,21 @@
...
 import org.springframework.data.domain.Page;
 import org.springframework.data.domain.PageImpl;
 import org.springframework.data.domain.PageRequest;
 import org.springframework.data.domain.Pageable;
 import org.springframework.data.domain.Sort;
 import org.springframework.data.util.Streamable;
 
+import com.fasterxml.jackson.annotation.JsonIgnore;
+
 public interface FoodieBaseService<T extends FoodieBaseEntity, ID> {
```
Comment (sfagade) [inline models/src/main/java/com/foodie/FoodieBaseService.java:19]: rename `ID`

File: models/src/main/java/com/foodie/FoodieBaseService.java [hunk 3]
```diff
@@ -86,4 +88,30 @@ default Page<T> findAllByDeletedAtIsNull(int page, int size, String sort) {
...
     }
-}
+
+    private void updateFields(T sourceEntity, T targetEntity) {
+        Class<?> clazz = sourceEntity.getClass();
+        Field[] fields = clazz.getDeclaredFields();
+        
+        for (Field field : fields) {
+            if (!isFieldIgnored(field)) {
+                field.setAccessible(true);
```
Comment (sfagade) [inline models/src/main/java/com/foodie/FoodieBaseService.java:98]: This violates the encapsulation principle

File: services/user-management/src/main/java/com/foodie/user/controllers/PermissionController.java [hunk 4]
```diff
@@ -1,18 +1,82 @@
...
+import java.util.Optional;
+import java.util.UUID;
+
 @Slf4j
 @RestController
 @CrossOrigin
 @RequestMapping("/api/v1/users/permissions")
 public class PermissionController {
     
     private PermissionService permissionService;
```
Comment (sfagade) [inline services/user-management/src/main/java/com/foodie/user/controllers/PermissionController.java:24]: This field can be final

File: services/user-management/src/main/java/com/foodie/user/controllers/PermissionController.java [hunk 5]
```diff
@@ -1,18 +1,82 @@
...
-}
+
+    public PermissionController(PermissionService permissionService) {
```
Comment (sfagade) [inline services/user-management/src/main/java/com/foodie/user/controllers/PermissionController.java:26]: I'd add `@Autowired` here to be safe

File: services/user-management/src/main/java/com/foodie/user/repositories/PermissionRepository.java [hunk 6]
```diff
@@ -1,15 +1,13 @@
...
 import com.foodie.user.model.Permission;
 
-import org.springframework.data.jpa.repository.JpaRepository;
-
 @Repository
-public interface PermissionRepository extends JpaRepository<Permission, UUID> {
-    Optional<Permission> findByPermissionName(String permissionName);
-}
+public interface PermissionRepository extends FoodieBaseRepository<Permission, UUID> {}
+    // Optional<Permission> findByPermissionName(String permissionName);
```
Comment (sfagade) [inline services/user-management/src/main/java/com/foodie/user/repositories/PermissionRepository.java:12]: I'm sure you meant to delete this?

File: services/user-management/src/test/java/com/foodie/user/controllers/PermissionControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+import org.springframework.http.ResponseEntity;
+
+import java.util.*;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.mockito.Mockito.*;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+@ExtendWith({ TestSetupPostgres.class })
+public class PermissionControllerTest {
```
Comment (sfagade) [inline services/user-management/src/test/java/com/foodie/user/controllers/PermissionControllerTest.java:25]: `public` modifier not needed here

**Review by sfagade — CHANGES_REQUESTED** [review]
Small and better

File: models/src/main/java/com/foodie/FoodieBaseService.java (hunk 2 above)
Comment (deye9) [reply models/src/main/java/com/foodie/FoodieBaseService.java:19]: Renamed to Key

File: models/src/main/java/com/foodie/FoodieBaseService.java (hunk 3 above)
Comment (deye9) [reply models/src/main/java/com/foodie/FoodieBaseService.java:98]: For now it is needed

File: services/user-management/src/main/java/com/foodie/user/controllers/PermissionController.java (hunk 4 above)
Comment (deye9) [reply services/user-management/src/main/java/com/foodie/user/controllers/PermissionController.java:24]: Making it final breaks the code.

File: services/user-management/src/main/java/com/foodie/user/repositories/PermissionRepository.java (hunk 6 above)
Comment (deye9) [reply services/user-management/src/main/java/com/foodie/user/repositories/PermissionRepository.java:12]: Thanks

File: models/src/main/java/com/foodie/FoodieBaseService.java (hunk 2 above)
Comment (deye9) [reply models/src/main/java/com/foodie/FoodieBaseService.java]: Nah Key is better AS I am dealing with the primary key

File: services/user-management/src/main/java/com/foodie/user/controllers/PermissionController.java (hunk 4 above)
Comment (sfagade) [reply services/user-management/src/main/java/com/foodie/user/controllers/PermissionController.java:24]: Screenshot 2024-03-25 at 9 10 52 pm
This should work

Comment (deye9) [reply services/user-management/src/main/java/com/foodie/user/controllers/PermissionController.java:24]: Final actually works. BReaks my test though but I will fix the test

File: models/src/main/java/com/foodie/FoodieBaseService.java (hunk 2 above)
Comment (sfagade) [reply models/src/main/java/com/foodie/FoodieBaseService.java]: As long as it's all lowercase `key`

File: services/user-management/src/main/java/com/foodie/user/controllers/PermissionController.java (hunk 4 above)
Comment (sfagade) [reply services/user-management/src/main/java/com/foodie/user/controllers/PermissionController.java:24]: If it breaks the test, then there's something else wrong with the code or the test

Comment (deye9) [reply services/user-management/src/main/java/com/foodie/user/controllers/PermissionController.java:24]: Issue is with the test

**Pull request merged**